<!DOCTYPE html>
<html lang="en"><head>
<script src="../../site_libs/clipboard/clipboard.min.js"></script>
<script src="../../site_libs/quarto-html/tabby.min.js"></script>
<script src="../../site_libs/quarto-html/popper.min.js"></script>
<script src="../../site_libs/quarto-html/tippy.umd.min.js"></script>
<link href="../../site_libs/quarto-html/tippy.css" rel="stylesheet">
<link href="../../site_libs/quarto-html/light-border.css" rel="stylesheet">
<link href="../../site_libs/quarto-html/quarto-syntax-highlighting-dark-8ea72dc5fed832574809a9c94082fbbb.css" rel="stylesheet" id="quarto-text-highlighting-styles"><meta charset="utf-8">
  <meta name="generator" content="quarto-1.6.40">

  <meta name="author" content="Eric R. Scott">
  <meta name="author" content="Kristina Riemer">
  <meta name="author" content="Renata Diaz">
  <meta name="dcterms.date" content="2024-06-20">
  <title>Intermediate ggplot2 workshop series – Exploring the Wide World of ggplot2 Extensions</title>
  <meta name="apple-mobile-web-app-capable" content="yes">
  <meta name="apple-mobile-web-app-status-bar-style" content="black-translucent">
  <meta name="viewport" content="width=device-width, initial-scale=1.0, maximum-scale=1.0, user-scalable=no, minimal-ui">
  <link rel="stylesheet" href="../../site_libs/revealjs/dist/reset.css">
  <link rel="stylesheet" href="../../site_libs/revealjs/dist/reveal.css">
  <style>
    code{white-space: pre-wrap;}
    span.smallcaps{font-variant: small-caps;}
    div.columns{display: flex; gap: min(4vw, 1.5em);}
    div.column{flex: auto; overflow-x: auto;}
    div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
    ul.task-list{list-style: none;}
    ul.task-list li input[type="checkbox"] {
      width: 0.8em;
      margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
      vertical-align: middle;
    }
    /* CSS for syntax highlighting */
    pre > code.sourceCode { white-space: pre; position: relative; }
    pre > code.sourceCode > span { line-height: 1.25; }
    pre > code.sourceCode > span:empty { height: 1.2em; }
    .sourceCode { overflow: visible; }
    code.sourceCode > span { color: inherit; text-decoration: inherit; }
    div.sourceCode { margin: 1em 0; }
    pre.sourceCode { margin: 0; }
    @media screen {
    div.sourceCode { overflow: auto; }
    }
    @media print {
    pre > code.sourceCode { white-space: pre-wrap; }
    pre > code.sourceCode > span { display: inline-block; text-indent: -5em; padding-left: 5em; }
    }
    pre.numberSource code
      { counter-reset: source-line 0; }
    pre.numberSource code > span
      { position: relative; left: -4em; counter-increment: source-line; }
    pre.numberSource code > span > a:first-child::before
      { content: counter(source-line);
        position: relative; left: -1em; text-align: right; vertical-align: baseline;
        border: none; display: inline-block;
        -webkit-touch-callout: none; -webkit-user-select: none;
        -khtml-user-select: none; -moz-user-select: none;
        -ms-user-select: none; user-select: none;
        padding: 0 4px; width: 4em;
      }
    pre.numberSource { margin-left: 3em;  padding-left: 4px; }
    div.sourceCode
      { color: #f8f8f2;  }
    @media screen {
    pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
    }
    code span { color: #f8f8f2; } /* Normal */
    code span.al { color: #f07178; background-color: #2a0f15; font-weight: bold; } /* Alert */
    code span.an { color: #d4d0ab; } /* Annotation */
    code span.at { color: #00e0e0; } /* Attribute */
    code span.bn { color: #d4d0ab; } /* BaseN */
    code span.bu { color: #abe338; } /* BuiltIn */
    code span.cf { color: #ffa07a; font-weight: bold; } /* ControlFlow */
    code span.ch { color: #abe338; } /* Char */
    code span.cn { color: #ffd700; } /* Constant */
    code span.co { color: #f8f8f2; font-style: italic; } /* Comment */
    code span.cv { color: #ffd700; } /* CommentVar */
    code span.do { color: #f8f8f2; } /* Documentation */
    code span.dt { color: #ffa07a; } /* DataType */
    code span.dv { color: #d4d0ab; } /* DecVal */
    code span.er { color: #f07178; text-decoration: underline; } /* Error */
    code span.ex { color: #00e0e0; font-weight: bold; } /* Extension */
    code span.fl { color: #d4d0ab; } /* Float */
    code span.fu { color: #ffa07a; } /* Function */
    code span.im { color: #abe338; } /* Import */
    code span.in { color: #d4d0ab; } /* Information */
    code span.kw { color: #ffa07a; font-weight: bold; } /* Keyword */
    code span.op { color: #ffa07a; } /* Operator */
    code span.ot { color: #00e0e0; } /* Other */
    code span.pp { color: #dcc6e0; } /* Preprocessor */
    code span.re { color: #00e0e0; background-color: #f8f8f2; } /* RegionMarker */
    code span.sc { color: #abe338; } /* SpecialChar */
    code span.ss { color: #abe338; } /* SpecialString */
    code span.st { color: #abe338; } /* String */
    code span.va { color: #00e0e0; } /* Variable */
    code span.vs { color: #abe338; } /* VerbatimString */
    code span.wa { color: #dcc6e0; } /* Warning */
  </style>
  <link rel="stylesheet" href="../../site_libs/revealjs/dist/theme/quarto-0a554b951af5837d684557d36d139718.css">
  <link href="../../site_libs/revealjs/plugin/quarto-line-highlight/line-highlight.css" rel="stylesheet">
  <link href="../../site_libs/revealjs/plugin/reveal-menu/menu.css" rel="stylesheet">
  <link href="../../site_libs/revealjs/plugin/reveal-menu/quarto-menu.css" rel="stylesheet">
  <link href="../../site_libs/revealjs/plugin/reveal-chalkboard/font-awesome/css/all.css" rel="stylesheet">
  <link href="../../site_libs/revealjs/plugin/reveal-chalkboard/style.css" rel="stylesheet">
  <link href="../../site_libs/revealjs/plugin/quarto-support/footer.css" rel="stylesheet">
  <style type="text/css">
    .reveal div.sourceCode {
      margin: 0;
      overflow: auto;
    }
    .reveal div.hanging-indent {
      margin-left: 1em;
      text-indent: -1em;
    }
    .reveal .slide:not(.center) {
      height: 100%;
    }
    .reveal .slide.scrollable {
      overflow-y: auto;
    }
    .reveal .footnotes {
      height: 100%;
      overflow-y: auto;
    }
    .reveal .slide .absolute {
      position: absolute;
      display: block;
    }
    .reveal .footnotes ol {
      counter-reset: ol;
      list-style-type: none; 
      margin-left: 0;
    }
    .reveal .footnotes ol li:before {
      counter-increment: ol;
      content: counter(ol) ". "; 
    }
    .reveal .footnotes ol li > p:first-child {
      display: inline-block;
    }
    .reveal .slide ul,
    .reveal .slide ol {
      margin-bottom: 0.5em;
    }
    .reveal .slide ul li,
    .reveal .slide ol li {
      margin-top: 0.4em;
      margin-bottom: 0.2em;
    }
    .reveal .slide ul[role="tablist"] li {
      margin-bottom: 0;
    }
    .reveal .slide ul li > *:first-child,
    .reveal .slide ol li > *:first-child {
      margin-block-start: 0;
    }
    .reveal .slide ul li > *:last-child,
    .reveal .slide ol li > *:last-child {
      margin-block-end: 0;
    }
    .reveal .slide .columns:nth-child(3) {
      margin-block-start: 0.8em;
    }
    .reveal blockquote {
      box-shadow: none;
    }
    .reveal .tippy-content>* {
      margin-top: 0.2em;
      margin-bottom: 0.7em;
    }
    .reveal .tippy-content>*:last-child {
      margin-bottom: 0.2em;
    }
    .reveal .slide > img.stretch.quarto-figure-center,
    .reveal .slide > img.r-stretch.quarto-figure-center {
      display: block;
      margin-left: auto;
      margin-right: auto; 
    }
    .reveal .slide > img.stretch.quarto-figure-left,
    .reveal .slide > img.r-stretch.quarto-figure-left  {
      display: block;
      margin-left: 0;
      margin-right: auto; 
    }
    .reveal .slide > img.stretch.quarto-figure-right,
    .reveal .slide > img.r-stretch.quarto-figure-right  {
      display: block;
      margin-left: auto;
      margin-right: 0; 
    }
  </style>
</head>
<body class="quarto-dark">
  <div class="reveal">
    <div class="slides">

<section id="title-slide" class="quarto-title-block center">
  <h1 class="title">Exploring the Wide World of ggplot2 Extensions</h1>

<div class="quarto-title-authors">
<div class="quarto-title-author">
<div class="quarto-title-author-name">
Eric R. Scott 
</div>
</div>
<div class="quarto-title-author">
<div class="quarto-title-author-name">
Kristina Riemer 
</div>
</div>
<div class="quarto-title-author">
<div class="quarto-title-author-name">
Renata Diaz 
</div>
</div>
</div>

  <p class="date">2024-06-20</p>
</section>
<section id="learning-objectives" class="slide level2">
<h2>Learning Objectives</h2>
<ul>
<li>Understand where to find packages that extend <code>ggplot2</code></li>
<li>Demonstrate some extension packages
<ul>
<li>Assemble multi-panel figures with <code>patchwork</code></li>
<li>Make animated plots with <code>gganimate</code></li>
<li>Visualize distributions with <code>ggdist</code></li>
<li>Plot network data with <code>ggraph</code></li>
</ul></li>
</ul>
</section>
<section id="packages" class="slide level2">
<h2>Packages</h2>
<div class="cell" data-layout-align="center">
<div class="sourceCode cell-code" id="cb1"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb1-1"><a></a><span class="fu">library</span>(ggplot2)</span>
<span id="cb1-2"><a></a><span class="fu">library</span>(palmerpenguins) <span class="co">#for example dataset</span></span>
<span id="cb1-3"><a></a><span class="fu">library</span>(patchwork) <span class="co">#for multi-panel figures</span></span>
<span id="cb1-4"><a></a><span class="fu">library</span>(gridGraphics) <span class="co">#for combining ggplot2 and base R figures</span></span>
<span id="cb1-5"><a></a><span class="fu">library</span>(gganimate) <span class="co">#for animated plots</span></span>
<span id="cb1-6"><a></a><span class="fu">library</span>(ggdist) <span class="co">#for showing distributions + uncertainty in data</span></span>
<span id="cb1-7"><a></a><span class="fu">library</span>(dplyr) <span class="co">#for cleaning data</span></span>
<span id="cb1-8"><a></a><span class="fu">library</span>(ggraph) <span class="co">#for network data</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section>
<section id="extensions-to-ggplot2" class="title-slide slide level1 center">
<h1>Extensions to <code>ggplot2</code></h1>
<p>There are many kinds of extensions to <code>ggplot2</code>, ranging from simple to complex, from familiar to transformative.</p>
</section>
<section id="general-use-and-modular" class="slide level2">
<h2>General-use and modular</h2>
<div class="cell" data-layout-align="center">
<div class="sourceCode cell-code" id="cb2" data-code-line-numbers="|1,4|2,5"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb2-1"><a></a><span class="fu">library</span>(ggbeeswarm)</span>
<span id="cb2-2"><a></a><span class="fu">library</span>(ggthemes)</span>
<span id="cb2-3"><a></a><span class="fu">ggplot</span>(penguins, <span class="fu">aes</span>(<span class="at">x =</span> species, <span class="at">y =</span> body_mass_g)) <span class="sc">+</span></span>
<span id="cb2-4"><a></a>  <span class="fu">geom_beeswarm</span>() <span class="sc">+</span></span>
<span id="cb2-5"><a></a>  <span class="fu">theme_economist</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>

</div>
<img data-src="slides_files/figure-revealjs/unnamed-chunk-3-1.png" class="quarto-figure quarto-figure-center r-stretch" width="960"><aside class="notes">
<p><code>geom_beeswarm()</code> can easily be combined with other <code>geom_</code>s</p>
<style type="text/css">
        span.MJX_Assistive_MathML {
          position:absolute!important;
          clip: rect(1px, 1px, 1px, 1px);
          padding: 1px 0 0 0!important;
          border: 0!important;
          height: 1px!important;
          width: 1px!important;
          overflow: hidden!important;
          display:block!important;
      }</style></aside>
</section>
<section id="all-in-one-functions" class="slide level2">
<h2>“All-in-one” functions</h2>
<div class="cell" data-layout-align="center">
<div class="sourceCode cell-code" id="cb3" data-code-line-numbers="|1,2|3"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb3-1"><a></a><span class="fu">library</span>(ggstatsplot)</span>
<span id="cb3-2"><a></a><span class="fu">ggbetweenstats</span>(penguins, species, body_mass_g) <span class="sc">+</span></span>
<span id="cb3-3"><a></a>  <span class="fu">scale_y_continuous</span>(<span class="st">"Body Mass (g)"</span>, <span class="at">n.breaks =</span> <span class="dv">10</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>

</div>
<img data-src="slides_files/figure-revealjs/unnamed-chunk-4-1.png" class="quarto-figure quarto-figure-center r-stretch" width="960"><aside class="notes">
<p>With great power comes great responsibility—make sure you trust these stats!<br>
Notice that the output is a ggplot object, so you can continue to add layers to it.</p>
<style type="text/css">
        span.MJX_Assistive_MathML {
          position:absolute!important;
          clip: rect(1px, 1px, 1px, 1px);
          padding: 1px 0 0 0!important;
          border: 0!important;
          height: 1px!important;
          width: 1px!important;
          overflow: hidden!important;
          display:block!important;
      }</style></aside>
</section>
<section id="transformative-field-specific" class="slide level2">
<h2>Transformative &amp; field-specific</h2>
<div>

</div>
<div class="quarto-layout-panel" data-layout-ncol="2">
<div class="quarto-layout-row">
<div class="quarto-layout-cell" style="flex-basis: 50.0%;justify-content: flex-start;">
<div class="quarto-figure quarto-figure-center">
<figure>
<p><a href="https://docs.r4photobiology.info/ggspectra"><img data-src="https://exts.ggplot2.tidyverse.org/gallery/images/ggspectra.png" style="width:100.0%" alt="ggspectra"></a></p>
<figcaption>ggspectra</figcaption>
</figure>
</div>
</div>
<div class="quarto-layout-cell" style="flex-basis: 50.0%;justify-content: flex-start;">
<div class="quarto-figure quarto-figure-center">
<figure>
<p><a href="https://wilkox.org/gggenes"><img data-src="https://exts.ggplot2.tidyverse.org/gallery/images/gggenes.png" style="width:100.0%" alt="gggenes"></a></p>
<figcaption>gggenes</figcaption>
</figure>
</div>
</div>
</div>
</div>
</section>
<section id="finding-extensions" class="slide level2">
<h2>Finding Extensions</h2>
<ul>
<li><p>Browse the <code>ggplot2</code> <a href="https://exts.ggplot2.tidyverse.org/gallery/">extensions gallery</a></p></li>
<li><p>Check out the <a href="https://github.com/erikgahner/awesome-ggplot2#readme">Awesome <code>ggplot2</code></a> list</p></li>
<li><p><a href="https://r-graph-gallery.com/">The R Graph Gallery</a></p></li>
<li><p>Google search with “ggplot2” keyword</p></li>
</ul>
</section></section>
<section>
<section id="package-demos" class="title-slide slide level1 center">
<h1>Package Demos</h1>

</section>
<section id="patchwork" class="slide level2">
<h2><code>patchwork</code></h2>
<p><code>patchwork</code> allows you to compose multi-panel figures with ease</p>

<img data-src="slides_files/figure-revealjs/unnamed-chunk-5-1.png" class="quarto-figure quarto-figure-center r-stretch" width="960"><aside class="notes">
<p>Things to notice:</p>
<ul>
<li>Plots areas are aligned despite one not having a y-axis label</li>
<li>Only one legend</li>
<li>Labels for panels A, B, C</li>
</ul>
<style type="text/css">
        span.MJX_Assistive_MathML {
          position:absolute!important;
          clip: rect(1px, 1px, 1px, 1px);
          padding: 1px 0 0 0!important;
          border: 0!important;
          height: 1px!important;
          width: 1px!important;
          overflow: hidden!important;
          display:block!important;
      }</style></aside>
</section>
<section id="example-plots" class="slide level2">
<h2>Example plots</h2>
<div class="cell" data-layout-align="center">
<div class="sourceCode cell-code" id="cb4"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb4-1"><a></a>p1 <span class="ot">&lt;-</span> </span>
<span id="cb4-2"><a></a>  <span class="fu">ggplot</span>(penguins, <span class="fu">aes</span>(<span class="at">x =</span> body_mass_g, </span>
<span id="cb4-3"><a></a>                       <span class="at">y =</span> bill_length_mm, </span>
<span id="cb4-4"><a></a>                       <span class="at">color =</span> species)) <span class="sc">+</span></span>
<span id="cb4-5"><a></a>  <span class="fu">geom_point</span>()</span>
<span id="cb4-6"><a></a>p2 <span class="ot">&lt;-</span></span>
<span id="cb4-7"><a></a>  <span class="fu">ggplot</span>(penguins, <span class="fu">aes</span>(<span class="at">x =</span> bill_depth_mm, </span>
<span id="cb4-8"><a></a>                       <span class="at">y =</span> bill_length_mm, </span>
<span id="cb4-9"><a></a>                       <span class="at">color =</span> species)) <span class="sc">+</span></span>
<span id="cb4-10"><a></a>  <span class="fu">geom_point</span>()</span>
<span id="cb4-11"><a></a>p3 <span class="ot">&lt;-</span> </span>
<span id="cb4-12"><a></a>  <span class="fu">ggplot</span>(penguins, <span class="fu">aes</span>(<span class="at">x =</span> body_mass_g)) <span class="sc">+</span> </span>
<span id="cb4-13"><a></a>  <span class="fu">facet_wrap</span>(<span class="fu">vars</span>(island)) <span class="sc">+</span></span>
<span id="cb4-14"><a></a>  <span class="fu">geom_histogram</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="combine-plots" class="slide level2">
<h2>Combine plots</h2>
<ul>
<li><code>+</code> wraps plots</li>
<li><code>|</code> combines plots horizontally</li>
<li><code>/</code> combines plots vertically</li>
<li><code>()</code> can be used to nest operations</li>
</ul>
</section>
<section id="combine-plots-1" class="slide level2">
<h2>Combine plots</h2>
<div class="cell" data-layout-align="center">
<div class="sourceCode cell-code" id="cb5"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb5-1"><a></a>(p1 <span class="sc">|</span> p2) <span class="sc">/</span> p3</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>

</div>
<img data-src="slides_files/figure-revealjs/unnamed-chunk-7-1.png" class="quarto-figure quarto-figure-center r-stretch" width="960"></section>
<section id="combine-guides" class="slide level2">
<h2>Combine guides</h2>
<p>If plots have <em>identical</em> guides, you can combine them with <code>plot_layout(guides = "collect")</code></p>
<div class="cell" data-layout-align="center">
<div class="sourceCode cell-code" id="cb6"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb6-1"><a></a>p1 <span class="sc">+</span> p2 <span class="sc">+</span> <span class="fu">plot_layout</span>(<span class="at">guides =</span> <span class="st">"collect"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>

</div>
<img data-src="slides_files/figure-revealjs/unnamed-chunk-8-1.png" class="quarto-figure quarto-figure-center r-stretch" width="960"><aside class="notes">
<p>identical means same scale, same title, same glyph</p>
<style type="text/css">
        span.MJX_Assistive_MathML {
          position:absolute!important;
          clip: rect(1px, 1px, 1px, 1px);
          padding: 1px 0 0 0!important;
          border: 0!important;
          height: 1px!important;
          width: 1px!important;
          overflow: hidden!important;
          display:block!important;
      }</style></aside>
</section>
<section id="combining-axes" class="slide level2">
<h2>Combining axes</h2>
<p>As of the most recent version of <code>patchwork</code> (v1.2.0) you can also combine identical axes.</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode cell-code" id="cb7"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb7-1"><a></a>p1 <span class="sc">+</span> p2 <span class="sc">+</span> <span class="fu">plot_layout</span>(<span class="at">guides =</span> <span class="st">"collect"</span>, <span class="at">axes =</span> <span class="st">"collect"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>

</div>
<img data-src="slides_files/figure-revealjs/unnamed-chunk-9-1.png" class="quarto-figure quarto-figure-center r-stretch" width="960"></section>
<section id="controlling-layout" class="slide level2">
<h2>Controlling layout</h2>
<p>For more options for controlling layout, see the related vignette on the package website.</p>
<ul>
<li>Adding empty areas</li>
<li>Inset plots</li>
<li>Adjusting widths and heights</li>
</ul>
</section>
<section id="tags" class="slide level2">
<h2>Tags</h2>
<p>You can add “tags” to each panel with <code>plot_annotation()</code></p>
<div class="cell" data-layout-align="center">
<div class="sourceCode cell-code" id="cb8"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb8-1"><a></a>p1 <span class="sc">+</span> p2 <span class="sc">+</span> <span class="fu">plot_annotation</span>(<span class="at">tag_levels =</span> <span class="st">"A"</span>, <span class="at">tag_suffix =</span> <span class="st">")"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>

</div>
<img data-src="slides_files/figure-revealjs/unnamed-chunk-10-1.png" class="quarto-figure quarto-figure-center r-stretch" width="960"></section>
<section id="modifying-all-panels" class="slide level2">
<h2>Modifying all panels</h2>
<p>You can use the <code>&amp;</code> operator instead of <code>+</code> to modify <strong>all</strong> elements of a multi-panel figure.</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode cell-code" id="cb9"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb9-1"><a></a>p1 <span class="sc">+</span> p2 <span class="sc">&amp;</span> <span class="fu">theme_bw</span>() <span class="sc">&amp;</span> <span class="fu">scale_color_viridis_d</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>

</div>
<img data-src="slides_files/figure-revealjs/unnamed-chunk-11-1.png" class="quarto-figure quarto-figure-center r-stretch" width="960"></section>
<section id="using-with-base-r-plots" class="slide level2">
<h2>Using with base R plots</h2>
<p>You can even combine <code>ggplot2</code> plots with base R plots with a special syntax</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode cell-code" id="cb10"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb10-1"><a></a>p1 <span class="sc">+</span> <span class="er">~</span><span class="fu">hist</span>(penguins<span class="sc">$</span>bill_depth_mm, <span class="at">main =</span> <span class="st">""</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>

</div>
<img data-src="slides_files/figure-revealjs/unnamed-chunk-12-1.png" class="quarto-figure quarto-figure-center r-stretch" width="960"></section>
<section id="gganimate" class="slide level2">
<h2><code>gganimate</code></h2>
<p>Display transitions between states of a variable</p>
<p>Useful for showing trends in time series &amp; spatial data</p>
<div class="callout callout-tip callout-titled callout-style-default">
<div class="callout-body">
<div class="callout-title">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<p><strong>Advice from the experts!</strong></p>
</div>
<div class="callout-content">
<p>“Graphic elements should only transition between instances of the same underlying phenomenon”</p>
</div>
</div>
</div>
</section>
<section id="example" class="slide level2">
<h2>Example</h2>
<p>Body size distribution of penguins changing over the years</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode cell-code" id="cb11"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb11-1"><a></a><span class="fu">ggplot</span>(penguins, <span class="fu">aes</span>(<span class="at">x =</span> body_mass_g, <span class="at">col =</span> species)) <span class="sc">+</span></span>
<span id="cb11-2"><a></a>  <span class="fu">geom_density</span>() <span class="sc">+</span></span>
<span id="cb11-3"><a></a>  <span class="fu">facet_wrap</span>(<span class="sc">~</span>year)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>

</div>
<img data-src="slides_files/figure-revealjs/unnamed-chunk-13-1.png" class="quarto-figure quarto-figure-center r-stretch" width="960"></section>
<section id="show-annual-transition" class="slide level2">
<h2>Show annual transition</h2>
<div class="cell" data-layout-align="center">
<div class="sourceCode cell-code" id="cb12"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb12-1"><a></a><span class="fu">library</span>(gganimate)</span>
<span id="cb12-2"><a></a></span>
<span id="cb12-3"><a></a><span class="fu">ggplot</span>(penguins, <span class="fu">aes</span>(<span class="at">x =</span> body_mass_g, <span class="at">col =</span> species)) <span class="sc">+</span></span>
<span id="cb12-4"><a></a>  <span class="fu">geom_density</span>() <span class="sc">+</span></span>
<span id="cb12-5"><a></a>  <span class="fu">transition_time</span>(year) <span class="sc">+</span></span>
<span id="cb12-6"><a></a>  <span class="fu">ggtitle</span>(<span class="st">'Year: {frame_time}'</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>

</div>
<img data-src="slides_files/figure-revealjs/unnamed-chunk-14-1.gif" class="quarto-figure quarto-figure-center r-stretch"></section>
<section id="add-history" class="slide level2">
<h2>Add history</h2>
<div class="cell" data-layout-align="center">
<div class="sourceCode cell-code" id="cb13"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb13-1"><a></a><span class="fu">ggplot</span>(penguins, <span class="fu">aes</span>(<span class="at">x =</span> body_mass_g, <span class="at">col =</span> species)) <span class="sc">+</span></span>
<span id="cb13-2"><a></a>  <span class="fu">geom_density</span>() <span class="sc">+</span></span>
<span id="cb13-3"><a></a>  <span class="fu">transition_time</span>(year) <span class="sc">+</span></span>
<span id="cb13-4"><a></a>  <span class="fu">ggtitle</span>(<span class="st">'Year: {frame_time}'</span>) <span class="sc">+</span></span>
<span id="cb13-5"><a></a>  <span class="fu">shadow_wake</span>(<span class="at">wake =</span> <span class="fl">0.3</span>, <span class="at">wrap =</span> <span class="cn">FALSE</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>

</div>
<img data-src="slides_files/figure-revealjs/unnamed-chunk-15-1.gif" class="quarto-figure quarto-figure-center r-stretch"></section>
<section id="modify-transition-speed" class="slide level2">
<h2>Modify transition speed</h2>
<div class="cell" data-layout-align="center">
<div class="sourceCode cell-code" id="cb14"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb14-1"><a></a><span class="fu">ggplot</span>(penguins, <span class="fu">aes</span>(<span class="at">x =</span> body_mass_g, <span class="at">col =</span> species)) <span class="sc">+</span></span>
<span id="cb14-2"><a></a>  <span class="fu">geom_density</span>() <span class="sc">+</span></span>
<span id="cb14-3"><a></a>  <span class="fu">transition_time</span>(year) <span class="sc">+</span></span>
<span id="cb14-4"><a></a>  <span class="fu">ggtitle</span>(<span class="st">'Year: {frame_time}'</span>) <span class="sc">+</span></span>
<span id="cb14-5"><a></a>  <span class="fu">ease_aes</span>(<span class="st">"quintic-in-out"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>

</div>
<img data-src="slides_files/figure-revealjs/unnamed-chunk-16-1.gif" class="quarto-figure quarto-figure-center r-stretch"></section>
<section id="have-axes-follow-data" class="slide level2">
<h2>Have axes follow data</h2>
<div class="cell" data-layout-align="center">
<div class="sourceCode cell-code" id="cb15"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb15-1"><a></a><span class="fu">ggplot</span>(penguins, <span class="fu">aes</span>(<span class="at">x =</span> body_mass_g, <span class="at">col =</span> species)) <span class="sc">+</span></span>
<span id="cb15-2"><a></a>  <span class="fu">geom_density</span>() <span class="sc">+</span></span>
<span id="cb15-3"><a></a>  <span class="fu">transition_time</span>(year) <span class="sc">+</span></span>
<span id="cb15-4"><a></a>  <span class="fu">ggtitle</span>(<span class="st">'Year: {frame_time}'</span>) <span class="sc">+</span></span>
<span id="cb15-5"><a></a>  <span class="fu">view_follow</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>

</div>
<img data-src="slides_files/figure-revealjs/unnamed-chunk-17-1.gif" class="quarto-figure quarto-figure-center r-stretch"></section>
<section id="save-file" class="slide level2">
<h2>Save file</h2>
<div class="cell" data-layout-align="center">
<div class="sourceCode cell-code" id="cb16"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb16-1"><a></a><span class="fu">anim_save</span>(<span class="st">"~/Desktop/test_anim.gif"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section></section>
<section>
<section id="ggdist-for-distributions" class="title-slide slide level1 smaller center">
<h1><code>ggdist</code> for distributions</h1>
<ul>
<li>Easily add nuance to density/histogram/ribbon plots</li>
<li>Particularly useful for:
<ul>
<li>Sample data</li>
<li>Fitted values + uncertainty</li>
<li>Frequentist or Bayesian parameter distributions</li>
</ul></li>
</ul>
</section>
<section id="slabinterval" class="slide level2">
<h2><code>slabinterval</code></h2>
<div class="cell" data-layout-align="center">
<div class="sourceCode cell-code" id="cb17"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb17-1"><a></a>penguins <span class="sc">%&gt;%</span></span>
<span id="cb17-2"><a></a>  <span class="fu">ggplot</span>(<span class="fu">aes</span>(<span class="at">x =</span> body_mass_g, <span class="at">y =</span> species)) <span class="sc">+</span></span>
<span id="cb17-3"><a></a>  <span class="fu">stat_slabinterval</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>

</div>
<img data-src="slides_files/figure-revealjs/slabinterval-1.png" class="quarto-figure quarto-figure-center r-stretch" width="960"><aside class="notes">
<ul>
<li>This plots a “halfeye” plot with a density distribution and a line at the bottom showing the mean and quantiles of the data.</li>
<li>Note that we didn’t have to create any new objects; stat_slabinterval automatically calculates densities and summary statistics. There are corresponding geom_* functions for already summarized data.</li>
<li>Other plot types within ggdist are variations on the stat_slabinterval theme.</li>
<li>Often you’ll want to use these shortcuts instead of building something up out of scratch.</li>
</ul>
<style type="text/css">
        span.MJX_Assistive_MathML {
          position:absolute!important;
          clip: rect(1px, 1px, 1px, 1px);
          padding: 1px 0 0 0!important;
          border: 0!important;
          height: 1px!important;
          width: 1px!important;
          overflow: hidden!important;
          display:block!important;
      }</style></aside>
</section>
<section id="example-dotsinterval" class="slide level2">
<h2>Example: <code>dotsinterval</code></h2>
<div class="cell" data-layout-align="center">
<div class="sourceCode cell-code" id="cb18"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb18-1"><a></a>penguins <span class="sc">%&gt;%</span></span>
<span id="cb18-2"><a></a>  <span class="fu">ggplot</span>(<span class="fu">aes</span>(<span class="at">x =</span> body_mass_g, <span class="at">y =</span> species)) <span class="sc">+</span></span>
<span id="cb18-3"><a></a>  <span class="fu">stat_dotsinterval</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>

</div>
<img data-src="slides_files/figure-revealjs/unnamed-chunk-20-1.png" class="quarto-figure quarto-figure-center r-stretch" width="960"></section>
<section id="example-interval" class="slide level2">
<h2>Example: <code>interval</code></h2>
<div class="cell" data-layout-align="center">
<div class="sourceCode cell-code" id="cb19"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb19-1"><a></a>penguins <span class="sc">%&gt;%</span></span>
<span id="cb19-2"><a></a>  <span class="fu">ggplot</span>(<span class="fu">aes</span>(<span class="at">x =</span> body_mass_g, <span class="at">y =</span> species)) <span class="sc">+</span></span>
<span id="cb19-3"><a></a>  <span class="fu">stat_interval</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>

</div>
<img data-src="slides_files/figure-revealjs/unnamed-chunk-21-1.png" class="quarto-figure quarto-figure-center r-stretch" width="960"></section>
<section id="example-interval-1" class="slide level2">
<h2>Example: <code>interval</code></h2>
<div class="cell" data-layout-align="center">
<div class="sourceCode cell-code" id="cb20"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb20-1"><a></a>penguins <span class="sc">%&gt;%</span></span>
<span id="cb20-2"><a></a>  <span class="fu">ggplot</span>(<span class="fu">aes</span>(<span class="at">x =</span> body_mass_g, <span class="at">y =</span> species)) <span class="sc">+</span></span>
<span id="cb20-3"><a></a>  <span class="fu">stat_interval</span>() <span class="sc">+</span></span>
<span id="cb20-4"><a></a>  <span class="fu">scale_color_brewer</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>

</div>
<img data-src="slides_files/figure-revealjs/unnamed-chunk-22-1.png" class="quarto-figure quarto-figure-center r-stretch" width="960"></section>
<section id="combining-elements-raincloud-plots" class="slide level2">
<h2>Combining elements: Raincloud plots</h2>
<div class="cell" data-layout-align="center">
<div class="sourceCode cell-code" id="cb21"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb21-1"><a></a>penguins <span class="sc">%&gt;%</span></span>
<span id="cb21-2"><a></a>  <span class="fu">ggplot</span>(<span class="fu">aes</span>(<span class="at">x =</span> body_mass_g, <span class="at">y =</span> species)) <span class="sc">+</span></span>
<span id="cb21-3"><a></a>  <span class="fu">stat_slabinterval</span>() <span class="sc">+</span></span>
<span id="cb21-4"><a></a>  <span class="fu">stat_dotsinterval</span>(<span class="at">side =</span> <span class="st">"bottom"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>

</div>
<img data-src="slides_files/figure-revealjs/unnamed-chunk-23-1.png" class="quarto-figure quarto-figure-center r-stretch" width="960"></section>
<section id="ribbon-plots" class="slide level2">
<h2>Ribbon plots</h2>
<p><code>stat_lineribbon</code> plots quantile intervals around a line automatically:</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode cell-code" id="cb22"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb22-1"><a></a>penguins <span class="sc">%&gt;%</span></span>
<span id="cb22-2"><a></a>  <span class="fu">ggplot</span>(<span class="fu">aes</span>(<span class="at">x =</span> year, <span class="at">y =</span> body_mass_g)) <span class="sc">+</span></span>
<span id="cb22-3"><a></a>  <span class="fu">stat_lineribbon</span>() <span class="sc">+</span></span>
<span id="cb22-4"><a></a>  <span class="fu">scale_fill_brewer</span>() <span class="sc">+</span></span>
<span id="cb22-5"><a></a>  <span class="fu">facet_wrap</span>(<span class="fu">vars</span>(species))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>

</div>
<img data-src="slides_files/figure-revealjs/unnamed-chunk-24-1.png" class="quarto-figure quarto-figure-center r-stretch" width="960"></section>
<section id="ribbon-plots-1" class="slide level2">
<h2>Ribbon plots</h2>
<p>You can control the bands using the <code>.width</code> argument:</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode cell-code" id="cb23"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb23-1"><a></a>penguins <span class="sc">%&gt;%</span></span>
<span id="cb23-2"><a></a>  <span class="fu">ggplot</span>(<span class="fu">aes</span>(<span class="at">x =</span> year, <span class="at">y =</span> body_mass_g)) <span class="sc">+</span></span>
<span id="cb23-3"><a></a>  <span class="fu">stat_lineribbon</span>(<span class="at">.width =</span> <span class="fu">c</span>(.<span class="dv">8</span>, .<span class="dv">97</span>, .<span class="dv">99</span>)) <span class="sc">+</span></span>
<span id="cb23-4"><a></a>  <span class="fu">scale_fill_brewer</span>() <span class="sc">+</span></span>
<span id="cb23-5"><a></a>  <span class="fu">facet_wrap</span>(<span class="fu">vars</span>(species))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>

</div>
<img data-src="slides_files/figure-revealjs/unnamed-chunk-25-1.png" class="quarto-figure quarto-figure-center r-stretch" width="960"></section>
<section id="ribbon-plots-2" class="slide level2">
<h2>Ribbon plots</h2>
<p><code>.width = ppoints()</code> creates a gradient:</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode cell-code" id="cb24"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb24-1"><a></a>penguins <span class="sc">%&gt;%</span></span>
<span id="cb24-2"><a></a>  <span class="fu">ggplot</span>(<span class="fu">aes</span>(<span class="at">x =</span> year, <span class="at">y =</span> body_mass_g, <span class="at">fill =</span> <span class="fu">after_stat</span>(.width))) <span class="sc">+</span></span>
<span id="cb24-3"><a></a>  <span class="fu">stat_lineribbon</span>(<span class="at">.width =</span> <span class="fu">ppoints</span>(<span class="dv">100</span>)) <span class="sc">+</span></span>
<span id="cb24-4"><a></a>  <span class="fu">scale_fill_distiller</span>() <span class="sc">+</span></span>
<span id="cb24-5"><a></a>  <span class="fu">facet_wrap</span>(<span class="fu">vars</span>(species))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>

</div>
<img data-src="slides_files/figure-revealjs/unnamed-chunk-26-1.png" class="quarto-figure quarto-figure-center r-stretch" width="960"></section>
<section id="visualizing-frequentist-model-output" class="slide level2">
<h2>Visualizing frequentist model output</h2>
<p>Combined with the <code>broom</code> and <code>distributional</code> packages, <code>ggdist</code> can display frequentist model uncertainty.</p>
</section>
<section id="visualizing-frequentist-model-output-1" class="slide level2">
<h2>Visualizing frequentist model output</h2>
<div class="cell" data-layout-align="center">
<div class="sourceCode cell-code" id="cb25"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb25-1"><a></a><span class="fu">library</span>(broom)</span>
<span id="cb25-2"><a></a><span class="fu">library</span>(distributional)</span>
<span id="cb25-3"><a></a></span>
<span id="cb25-4"><a></a>penguin_lm <span class="ot">&lt;-</span> <span class="fu">lm</span>(body_mass_g <span class="sc">~</span>  species, <span class="at">data =</span> penguins)</span>
<span id="cb25-5"><a></a></span>
<span id="cb25-6"><a></a>broom<span class="sc">::</span><span class="fu">tidy</span>(penguin_lm)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 3 × 5
  term             estimate std.error statistic   p.value
  &lt;chr&gt;               &lt;dbl&gt;     &lt;dbl&gt;     &lt;dbl&gt;     &lt;dbl&gt;
1 (Intercept)        3706.       38.1    97.2   6.88e-245
2 speciesChinstrap     26.9      67.7     0.398 6.91e-  1
3 speciesGentoo      1386.       56.9    24.4   1.01e- 75</code></pre>
</div>
</div>
</section>
<section id="visualizing-frequentist-model-output-2" class="slide level2">
<h2>Visualizing frequentist model output</h2>
<div class="cell" data-layout-align="center">
<div class="sourceCode cell-code" id="cb27"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb27-1"><a></a>broom<span class="sc">::</span><span class="fu">tidy</span>(penguin_lm) <span class="sc">|&gt;</span></span>
<span id="cb27-2"><a></a>  <span class="fu">ggplot</span>(<span class="fu">aes</span>(<span class="at">y =</span> term)) <span class="sc">+</span></span>
<span id="cb27-3"><a></a>    <span class="fu">stat_halfeye</span>(</span>
<span id="cb27-4"><a></a>      <span class="fu">aes</span>(<span class="at">xdist =</span> <span class="fu">dist_student_t</span>(<span class="at">df =</span> <span class="fu">df.residual</span>(penguin_lm), </span>
<span id="cb27-5"><a></a>                                 <span class="at">mu =</span> estimate, </span>
<span id="cb27-6"><a></a>                                 <span class="at">sigma =</span> std.error))</span>
<span id="cb27-7"><a></a>    )</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>

</div>
<img data-src="slides_files/figure-revealjs/unnamed-chunk-28-1.png" class="quarto-figure quarto-figure-center r-stretch" width="960"></section>
<section id="visualizing-frequentist-model-output-3" class="slide level2">
<h2>Visualizing frequentist model output</h2>
<ul>
<li>You can also use ribbons to show uncertainty around lines of fit.</li>
<li>We’ll need a linear model with a continuous predictor.</li>
<li>And we’ll use <code>tidyr::expand</code> and <code>broom::augment</code> to get the predicted line of fit from the model.</li>
</ul>
</section>
<section id="visualizing-frequentist-model-output-4" class="slide level2">
<h2>Visualizing frequentist model output</h2>
<div class="cell" data-layout-align="center">
<div class="sourceCode cell-code" id="cb28"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb28-1"><a></a>penguin_lm_cont <span class="ot">&lt;-</span> <span class="fu">lm</span>(body_mass_g <span class="sc">~</span>  year <span class="sc">+</span> species, <span class="at">data =</span> penguins)</span>
<span id="cb28-2"><a></a></span>
<span id="cb28-3"><a></a>penguin_lm_cont_fitted <span class="ot">&lt;-</span> penguins <span class="sc">%&gt;%</span></span>
<span id="cb28-4"><a></a>  <span class="fu">group_by</span>(species) <span class="sc">%&gt;%</span></span>
<span id="cb28-5"><a></a>  tidyr<span class="sc">::</span><span class="fu">expand</span>(<span class="at">year =</span> <span class="fu">seq</span>(<span class="fu">min</span>(year), <span class="fu">max</span>(year), <span class="at">length.out =</span> <span class="dv">3</span>)) <span class="sc">%&gt;%</span></span>
<span id="cb28-6"><a></a>  <span class="fu">augment</span>(penguin_lm_cont, <span class="at">newdata =</span> ., <span class="at">se_fit =</span> <span class="cn">TRUE</span>) </span>
<span id="cb28-7"><a></a></span>
<span id="cb28-8"><a></a><span class="fu">head</span>(penguin_lm_cont_fitted) </span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 6 × 4
  species    year .fitted .se.fit
  &lt;fct&gt;     &lt;dbl&gt;   &lt;dbl&gt;   &lt;dbl&gt;
1 Adelie     2007   3703.    50.4
2 Adelie     2008   3706.    38.2
3 Adelie     2009   3709.    48.2
4 Chinstrap  2007   3730.    63.6
5 Chinstrap  2008   3733.    56.0
6 Chinstrap  2009   3737.    64.5</code></pre>
</div>
</div>
</section>
<section id="visualizing-frequentist-model-output-5" class="slide level2">
<h2>Visualizing frequentist model output</h2>
<div class="cell" data-layout-align="center">
<div class="sourceCode cell-code" id="cb30"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb30-1"><a></a>  <span class="fu">ggplot</span>(penguin_lm_cont_fitted, <span class="fu">aes</span>(<span class="at">x =</span> year)) <span class="sc">+</span></span>
<span id="cb30-2"><a></a>  <span class="fu">stat_lineribbon</span>(</span>
<span id="cb30-3"><a></a>    <span class="fu">aes</span>(<span class="at">ydist =</span> <span class="fu">dist_student_t</span>(<span class="at">df =</span> <span class="fu">df.residual</span>(penguin_lm_cont), </span>
<span id="cb30-4"><a></a>                               <span class="at">mu =</span> .fitted, </span>
<span id="cb30-5"><a></a>                               <span class="at">sigma =</span> .se.fit))) <span class="sc">+</span></span>
<span id="cb30-6"><a></a>  <span class="fu">facet_wrap</span>(<span class="fu">vars</span>(species)) <span class="sc">+</span></span>
<span id="cb30-7"><a></a>  <span class="fu">scale_fill_brewer</span>() </span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>

</div>
<img data-src="slides_files/figure-revealjs/unnamed-chunk-30-1.png" class="quarto-figure quarto-figure-center r-stretch" width="960"></section>
<section id="visualizing-bayesian-model-components" class="slide level2">
<h2>Visualizing Bayesian model components</h2>
<ul>
<li><code>ggdist</code> plots are great for visualizing Bayesian model components:
<ul>
<li>priors</li>
<li>posterior draws</li>
<li>posterior predictions</li>
</ul></li>
<li>We can fit a simple Bayesian linear model and visualize the posterior…</li>
</ul>
</section>
<section id="fitting-a-model" class="slide level2">
<h2>Fitting a model</h2>
<div class="cell" data-layout-align="center">
<div class="sourceCode cell-code" id="cb31"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb31-1"><a></a><span class="fu">library</span>(brms)</span>
<span id="cb31-2"><a></a><span class="fu">library</span>(tidybayes)</span>
<span id="cb31-3"><a></a></span>
<span id="cb31-4"><a></a>penguins_brm <span class="ot">&lt;-</span> <span class="fu">brm</span>(body_mass_g <span class="sc">~</span> species, <span class="at">data =</span> penguins, <span class="at">iter =</span> <span class="dv">1000</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="extracting-draws-from-the-posterior" class="slide level2">
<h2>Extracting draws from the posterior</h2>
<div class="cell" data-layout-align="center">
<div class="sourceCode cell-code" id="cb32"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb32-1"><a></a>penguins_brm <span class="sc">%&gt;%</span></span>
<span id="cb32-2"><a></a>tidybayes<span class="sc">::</span><span class="fu">gather_draws</span>(</span>
<span id="cb32-3"><a></a><span class="fu">c</span>(b_Intercept,</span>
<span id="cb32-4"><a></a>b_speciesChinstrap,</span>
<span id="cb32-5"><a></a>b_speciesGentoo)</span>
<span id="cb32-6"><a></a>) <span class="sc">|&gt;</span></span>
<span id="cb32-7"><a></a><span class="fu">head</span>() </span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 6 × 5
# Groups:   .variable [1]
  .chain .iteration .draw .variable   .value
   &lt;int&gt;      &lt;int&gt; &lt;int&gt; &lt;chr&gt;        &lt;dbl&gt;
1      1          1     1 b_Intercept  3646.
2      1          2     2 b_Intercept  3680.
3      1          3     3 b_Intercept  3694.
4      1          4     4 b_Intercept  3688.
5      1          5     5 b_Intercept  3632.
6      1          6     6 b_Intercept  3779.</code></pre>
</div>
</div>
</section>
<section id="posterior-draws-rainclouds" class="slide level2">
<h2>Posterior draws rainclouds</h2>
<div class="cell" data-layout-align="center">
<div class="sourceCode cell-code" id="cb34"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb34-1"><a></a>penguins_brm <span class="sc">|&gt;</span></span>
<span id="cb34-2"><a></a>tidybayes<span class="sc">::</span><span class="fu">gather_draws</span>(</span>
<span id="cb34-3"><a></a><span class="fu">c</span>(b_Intercept,</span>
<span id="cb34-4"><a></a>b_speciesChinstrap,</span>
<span id="cb34-5"><a></a>b_speciesGentoo)) <span class="sc">%&gt;%</span></span>
<span id="cb34-6"><a></a><span class="fu">ggplot</span>(<span class="fu">aes</span>(<span class="at">y =</span> .variable, <span class="at">x =</span> .value)) <span class="sc">+</span>  </span>
<span id="cb34-7"><a></a>  <span class="fu">geom_dotsinterval</span>(<span class="at">side =</span> <span class="st">"bottom"</span>, <span class="at">dotsize =</span> .<span class="dv">05</span>) <span class="sc">+</span> </span>
<span id="cb34-8"><a></a>  <span class="fu">stat_slabinterval</span>() </span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>

</div>
<img data-src="slides_files/figure-revealjs/unnamed-chunk-34-1.png" class="quarto-figure quarto-figure-center r-stretch" width="960"></section>
<section id="takeaways" class="slide level2">
<h2>Takeaways</h2>
<ul>
<li><code>ggdist</code> contains <code>geom</code> and <code>stat</code> functions for plotting distributions.</li>
<li>Slab functions show variations on density plots.</li>
<li>Ribbons show variation around a line.</li>
<li><code>ggdist</code> plays nicely with Bayesian and frequentist modeling frameworks.</li>
</ul>
</section></section>
<section>
<section id="relational-data-with-ggraph" class="title-slide slide level1 center">
<h1>Relational data with <code>ggraph</code></h1>

</section>
<section id="ggraph" class="slide level2 smaller">
<h2><code>ggraph</code></h2>
<ul>
<li>For plotting networks, graphs, and trees</li>
<li>Adds <strong>layouts</strong> and geoms for nodes and edges</li>
<li>Integrates with <a href="https://ggraph.data-imaginist.com/articles/tidygraph.html"><code>tidygraph</code></a> to wrap around numerous other graph object types (<code>igraph</code>, <code>dedrogram</code>, <code>hclust</code>, <code>graph</code>, <code>phylo</code>, …)</li>
</ul>
</section>
<section id="example-mouse-data" class="slide level2">
<h2>Example: Mouse data</h2>
<p>Begin by loading network data, in this case from the <code>graphml</code> format.</p>
<p>Data from <a href="https://journals.plos.org/plosone/article?id=10.1371/journal.pone.0134509">So et al.&nbsp;2015</a>, in the <a href="https://github.com/bansallab/asnr/tree/master">Animal Social Networks Repository</a>.</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode cell-code" id="cb35"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb35-1"><a></a><span class="fu">library</span>(igraph)</span>
<span id="cb35-2"><a></a><span class="fu">library</span>(tidygraph)</span>
<span id="cb35-3"><a></a></span>
<span id="cb35-4"><a></a>mouse_sniffing <span class="ot">&lt;-</span> <span class="fu">read_graph</span>(here<span class="sc">::</span><span class="fu">here</span>(<span class="st">"2024"</span>, <span class="st">"03-extensions"</span>, <span class="st">"mouse_so_grooming_network.graphml"</span>),</span>
<span id="cb35-5"><a></a><span class="at">format =</span> <span class="st">"graphml"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="example-as_tbl_graph" class="slide level2">
<h2>Example: <code>as_tbl_graph</code></h2>
<p><code>tidygraph::as_tbl_graph</code> standardizes different graph storage formats.</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode cell-code" id="cb36"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb36-1"><a></a>mouse_graph <span class="ot">&lt;-</span> mouse_sniffing <span class="sc">|&gt;</span></span>
<span id="cb36-2"><a></a><span class="fu">as_tbl_graph</span>() </span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="plotting-mouse-sniffing-data" class="slide level2">
<h2>Plotting mouse sniffing data</h2>
<div class="cell" data-layout-align="center">
<div class="sourceCode cell-code" id="cb37"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb37-1"><a></a><span class="fu">ggraph</span>(mouse_graph, <span class="at">layout =</span> <span class="st">'kk'</span>) <span class="sc">+</span> </span>
<span id="cb37-2"><a></a>    <span class="fu">geom_edge_fan</span>() <span class="sc">+</span> </span>
<span id="cb37-3"><a></a>    <span class="fu">geom_node_point</span>() <span class="sc">+</span></span>
<span id="cb37-4"><a></a>    <span class="fu">theme_minimal</span>() <span class="sc">+</span></span>
<span id="cb37-5"><a></a>    <span class="fu">ggtitle</span>(<span class="st">"Mouse sniffs"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>

</div>
<img data-src="slides_files/figure-revealjs/unnamed-chunk-37-1.png" class="quarto-figure quarto-figure-center r-stretch" width="960"></section>
<section id="alternate-layouts" class="slide level2">
<h2>Alternate layouts</h2>
<div class="cell" data-layout-align="center">
<div class="sourceCode cell-code" id="cb38"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb38-1"><a></a><span class="fu">ggraph</span>(mouse_graph, <span class="at">layout =</span> <span class="st">'eigen'</span>) <span class="sc">+</span> </span>
<span id="cb38-2"><a></a>    <span class="fu">geom_edge_fan</span>() <span class="sc">+</span> </span>
<span id="cb38-3"><a></a>    <span class="fu">geom_node_point</span>() <span class="sc">+</span></span>
<span id="cb38-4"><a></a>    <span class="fu">theme_minimal</span>() <span class="sc">+</span></span>
<span id="cb38-5"><a></a>    <span class="fu">ggtitle</span>(<span class="st">"Mouse sniffs"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>

</div>
<img data-src="slides_files/figure-revealjs/unnamed-chunk-38-1.png" class="quarto-figure quarto-figure-center r-stretch" width="960"></section>
<section id="alternate-layouts-1" class="slide level2">
<h2>Alternate layouts</h2>
<div class="cell" data-layout-align="center">
<div class="sourceCode cell-code" id="cb39"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb39-1"><a></a><span class="fu">ggraph</span>(mouse_graph, <span class="at">layout =</span> <span class="st">'linear'</span>, <span class="at">circular =</span> <span class="cn">TRUE</span>) <span class="sc">+</span> </span>
<span id="cb39-2"><a></a>    <span class="fu">geom_edge_fan</span>() <span class="sc">+</span> </span>
<span id="cb39-3"><a></a>    <span class="fu">geom_node_point</span>() <span class="sc">+</span></span>
<span id="cb39-4"><a></a>    <span class="fu">theme_minimal</span>() <span class="sc">+</span></span>
<span id="cb39-5"><a></a>    <span class="fu">ggtitle</span>(<span class="st">"Mouse sniffs"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>

</div>
<img data-src="slides_files/figure-revealjs/unnamed-chunk-39-1.png" class="quarto-figure quarto-figure-center r-stretch" width="960"></section>
<section id="example-phylogenies" class="slide level2">
<h2>Example: phylogenies</h2>
<ul>
<li><code>ggraph</code> can plot phylogenetic trees.</li>
<li>We’ll get a phylogeny using the R Open Tree of Life (<code>rotl</code>) and plot it.</li>
</ul>
</section>
<section id="getting-a-phylogeny" class="slide level2">
<h2>Getting a phylogeny</h2>

<img data-src="krat.jpg" class="quarto-figure quarto-figure-center r-stretch"></section>
<section id="getting-a-phylogeny-1" class="slide level2">
<h2>Getting a phylogeny</h2>
<div class="cell" data-layout-align="center">
<div class="sourceCode cell-code" id="cb40"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb40-1"><a></a><span class="fu">library</span>(rotl)</span>
<span id="cb40-2"><a></a>rodent_id <span class="ot">&lt;-</span> <span class="fu">tnrs_match_names</span>(<span class="st">"Heteromyidae"</span>)</span>
<span id="cb40-3"><a></a>rodent_tree <span class="ot">&lt;-</span> <span class="fu">tol_subtree</span>(<span class="at">ott_id =</span> <span class="fu">ott_id</span>(rodent_id))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>
Progress [----------------------------------] 0/11 (  0) ?s
Progress [================================] 11/11 (100)  0s
                                                            </code></pre>
</div>
<div class="sourceCode cell-code" id="cb42"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb42-1"><a></a>rodent_tree<span class="sc">$</span>node.label <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="dv">1</span><span class="sc">:</span><span class="dv">64</span>)</span>
<span id="cb42-2"><a></a><span class="fu">class</span>(rodent_tree)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] "phylo"</code></pre>
</div>
</div>
</section>
<section id="convert-the-phylogeny-to-a-graph" class="slide level2">
<h2>Convert the phylogeny to a graph</h2>
<div class="cell" data-layout-align="center">
<div class="sourceCode cell-code" id="cb44"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb44-1"><a></a>rodent_graph <span class="ot">&lt;-</span> <span class="fu">as_tbl_graph</span>(rodent_tree)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="dendrogram" class="slide level2">
<h2>Dendrogram</h2>
<div class="cell" data-layout-align="center">
<div class="sourceCode cell-code" id="cb45"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb45-1"><a></a><span class="fu">ggraph</span>(rodent_graph, <span class="st">"dendrogram"</span>) <span class="sc">+</span></span>
<span id="cb45-2"><a></a><span class="fu">geom_edge_link</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>

</div>
<img data-src="slides_files/figure-revealjs/unnamed-chunk-42-1.png" class="quarto-figure quarto-figure-center r-stretch" width="960"></section>
<section id="unrooted-tree" class="slide level2">
<h2>Unrooted tree</h2>
<div class="cell" data-layout-align="center">
<div class="sourceCode cell-code" id="cb46"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb46-1"><a></a><span class="fu">ggraph</span>(rodent_graph, <span class="st">"unrooted"</span>) <span class="sc">+</span></span>
<span id="cb46-2"><a></a><span class="fu">geom_edge_link</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>

</div>
<img data-src="slides_files/figure-revealjs/unnamed-chunk-43-1.png" class="quarto-figure quarto-figure-center r-stretch" width="960"></section>
<section id="takeaways-1" class="slide level2">
<h2>Takeaways</h2>
<ul>
<li><code>ggraph</code> can plot many different types of network objects.</li>
<li>Use <code>as_tbl_graph</code> to convert graphs to <code>ggraph</code>-compatible objects.</li>
<li>Layouts strongly determine the appearance and interpretability of a graph.</li>
</ul>
</section>
<section id="extensions-resources" class="slide level2">
<h2>Extensions Resources</h2>
<p><code>patchwork</code></p>
<ul>
<li><a href="https://patchwork.data-imaginist.com/index.html">Package website</a></li>
</ul>
<p><code>gganimate</code></p>
<ul>
<li><a href="https://rstudio.github.io/cheatsheets/gganimate.pdf">Cheat sheet</a></li>
<li><a href="https://gganimate.com">Website</a></li>
</ul>
</section>
<section id="extensions-resources-1" class="slide level2">
<h2>Extensions Resources</h2>
<p><code>ggdist</code></p>
<ul>
<li><a href="https://mjskay.github.io/ggdist/index.html">Documentation</a></li>
<li>More about <a href="http://mjskay.github.io/tidybayes/">tidybayes</a></li>
</ul>
<p><code>ggraph</code></p>
<ul>
<li><a href="https://ggraph.data-imaginist.com/index.html"><code>ggraph</code> documentation</a></li>
<li><a href="https://tidygraph.data-imaginist.com/"><code>tidygraph</code> documentation</a></li>
<li><a href="https://r.igraph.org/"><code>igraph</code> R package documentation</a></li>
<li><a href="https://cran.r-project.org/web/packages/rotl/rotl.pdf"><code>rotl</code> documentation</a></li>
</ul>
</section>
<section id="getting-help" class="slide level2">
<h2>Getting Help</h2>
<ul>
<li><p>Our <a href="https://datascience.cct.arizona.edu/drop-in-hours">drop-in hours</a></p></li>
<li><p><a href="https://join.slack.com/t/uadatascience/shared_invite/zt-1xjxht9k0-E489WA6axO_SRVeKQh3cXg">UA Data Science Slack</a></p></li>
</ul>
</section></section>
<section>
<section id="anything-we-missed" class="title-slide slide level1 center">
<h1>Anything we missed?</h1>
<p>Got a <code>ggplot2</code> question we didn’t cover in this workshop series? Let’s figure it out together!</p>
</section>
<section id="data-sources" class="slide level2">
<h2>Data sources</h2>
<p>Pratha Sah, José David Mendez, Shweta Bansal. A multi-species repository of social networks. Scientific Data, 6:44 (2019)</p>
<p>So, N., Franks, B., Lim, S., and Curley, J.P. 2015. A social network approach reveals associations between mouse social dominance and brain gene expression. PlosOne 10(7):e0134509</p>


</section></section>
    </div>
  <div class="quarto-auto-generated-content" style="display: none;">
<p><img src="../../logo.png" class="slide-logo"></p>
<div class="footer footer-default">
<p><a href="../../2024/03-extensions/index.html">Intermediate ggplot2 workshop series</a></p>
</div>
</div></div>

  <script>window.backupDefine = window.define; window.define = undefined;</script>
  <script src="../../site_libs/revealjs/dist/reveal.js"></script>
  <!-- reveal.js plugins -->
  <script src="../../site_libs/revealjs/plugin/quarto-line-highlight/line-highlight.js"></script>
  <script src="../../site_libs/revealjs/plugin/pdf-export/pdfexport.js"></script>
  <script src="../../site_libs/revealjs/plugin/reveal-menu/menu.js"></script>
  <script src="../../site_libs/revealjs/plugin/reveal-menu/quarto-menu.js"></script>
  <script src="../../site_libs/revealjs/plugin/reveal-chalkboard/plugin.js"></script>
  <script src="../../site_libs/revealjs/plugin/quarto-support/support.js"></script>
  

  <script src="../../site_libs/revealjs/plugin/notes/notes.js"></script>
  <script src="../../site_libs/revealjs/plugin/search/search.js"></script>
  <script src="../../site_libs/revealjs/plugin/zoom/zoom.js"></script>
  <script src="../../site_libs/revealjs/plugin/math/math.js"></script>
  <script>window.define = window.backupDefine; window.backupDefine = undefined;</script>

  <script>

      // Full list of configuration options available at:
      // https://revealjs.com/config/
      Reveal.initialize({
'controlsAuto': true,
'previewLinksAuto': false,
'pdfSeparateFragments': false,
'autoAnimateEasing': "ease",
'autoAnimateDuration': 1,
'autoAnimateUnmatched': true,
'jumpToSlide': true,
'menu': {"side":"left","useTextContentForMissingTitles":true,"markers":false,"loadIcons":false,"custom":[{"title":"Tools","icon":"<i class=\"fas fa-gear\"></i>","content":"<ul class=\"slide-menu-items\">\n<li class=\"slide-tool-item active\" data-item=\"0\"><a href=\"#\" onclick=\"RevealMenuToolHandlers.fullscreen(event)\"><kbd>f</kbd> Fullscreen</a></li>\n<li class=\"slide-tool-item\" data-item=\"1\"><a href=\"#\" onclick=\"RevealMenuToolHandlers.speakerMode(event)\"><kbd>s</kbd> Speaker View</a></li>\n<li class=\"slide-tool-item\" data-item=\"2\"><a href=\"#\" onclick=\"RevealMenuToolHandlers.overview(event)\"><kbd>o</kbd> Slide Overview</a></li>\n<li class=\"slide-tool-item\" data-item=\"3\"><a href=\"#\" onclick=\"RevealMenuToolHandlers.togglePdfExport(event)\"><kbd>e</kbd> PDF Export Mode</a></li>\n<li class=\"slide-tool-item\" data-item=\"4\"><a href=\"#\" onclick=\"RevealMenuToolHandlers.toggleScrollView(event)\"><kbd>r</kbd> Scroll View Mode</a></li>\n<li class=\"slide-tool-item\" data-item=\"5\"><a href=\"#\" onclick=\"RevealMenuToolHandlers.toggleChalkboard(event)\"><kbd>b</kbd> Toggle Chalkboard</a></li>\n<li class=\"slide-tool-item\" data-item=\"6\"><a href=\"#\" onclick=\"RevealMenuToolHandlers.toggleNotesCanvas(event)\"><kbd>c</kbd> Toggle Notes Canvas</a></li>\n<li class=\"slide-tool-item\" data-item=\"7\"><a href=\"#\" onclick=\"RevealMenuToolHandlers.downloadDrawings(event)\"><kbd>d</kbd> Download Drawings</a></li>\n<li class=\"slide-tool-item\" data-item=\"8\"><a href=\"#\" onclick=\"RevealMenuToolHandlers.keyboardHelp(event)\"><kbd>?</kbd> Keyboard Help</a></li>\n</ul>"}],"openButton":true},
'chalkboard': {"buttons":true},
'smaller': false,
 
        // Display controls in the bottom right corner
        controls: false,

        // Help the user learn the controls by providing hints, for example by
        // bouncing the down arrow when they first encounter a vertical slide
        controlsTutorial: false,

        // Determines where controls appear, "edges" or "bottom-right"
        controlsLayout: 'edges',

        // Visibility rule for backwards navigation arrows; "faded", "hidden"
        // or "visible"
        controlsBackArrows: 'faded',

        // Display a presentation progress bar
        progress: true,

        // Display the page number of the current slide
        slideNumber: 'c/t',

        // 'all', 'print', or 'speaker'
        showSlideNumber: 'all',

        // Add the current slide number to the URL hash so that reloading the
        // page/copying the URL will return you to the same slide
        hash: true,

        // Start with 1 for the hash rather than 0
        hashOneBasedIndex: false,

        // Flags if we should monitor the hash and change slides accordingly
        respondToHashChanges: true,

        // Push each slide change to the browser history
        history: true,

        // Enable keyboard shortcuts for navigation
        keyboard: true,

        // Enable the slide overview mode
        overview: true,

        // Disables the default reveal.js slide layout (scaling and centering)
        // so that you can use custom CSS layout
        disableLayout: false,

        // Vertical centering of slides
        center: false,

        // Enables touch navigation on devices with touch input
        touch: true,

        // Loop the presentation
        loop: false,

        // Change the presentation direction to be RTL
        rtl: false,

        // see https://revealjs.com/vertical-slides/#navigation-mode
        navigationMode: 'linear',

        // Randomizes the order of slides each time the presentation loads
        shuffle: false,

        // Turns fragments on and off globally
        fragments: true,

        // Flags whether to include the current fragment in the URL,
        // so that reloading brings you to the same fragment position
        fragmentInURL: false,

        // Flags if the presentation is running in an embedded mode,
        // i.e. contained within a limited portion of the screen
        embedded: false,

        // Flags if we should show a help overlay when the questionmark
        // key is pressed
        help: true,

        // Flags if it should be possible to pause the presentation (blackout)
        pause: true,

        // Flags if speaker notes should be visible to all viewers
        showNotes: false,

        // Global override for autoplaying embedded media (null/true/false)
        autoPlayMedia: null,

        // Global override for preloading lazy-loaded iframes (null/true/false)
        preloadIframes: null,

        // Number of milliseconds between automatically proceeding to the
        // next slide, disabled when set to 0, this value can be overwritten
        // by using a data-autoslide attribute on your slides
        autoSlide: 0,

        // Stop auto-sliding after user input
        autoSlideStoppable: true,

        // Use this method for navigation when auto-sliding
        autoSlideMethod: null,

        // Specify the average time in seconds that you think you will spend
        // presenting each slide. This is used to show a pacing timer in the
        // speaker view
        defaultTiming: null,

        // Enable slide navigation via mouse wheel
        mouseWheel: false,

        // The display mode that will be used to show slides
        display: 'block',

        // Hide cursor if inactive
        hideInactiveCursor: true,

        // Time before the cursor is hidden (in ms)
        hideCursorTime: 5000,

        // Opens links in an iframe preview overlay
        previewLinks: false,

        // Transition style (none/fade/slide/convex/concave/zoom)
        transition: 'none',

        // Transition speed (default/fast/slow)
        transitionSpeed: 'default',

        // Transition style for full page slide backgrounds
        // (none/fade/slide/convex/concave/zoom)
        backgroundTransition: 'none',

        // Number of slides away from the current that are visible
        viewDistance: 3,

        // Number of slides away from the current that are visible on mobile
        // devices. It is advisable to set this to a lower number than
        // viewDistance in order to save resources.
        mobileViewDistance: 2,

        // The "normal" size of the presentation, aspect ratio will be preserved
        // when the presentation is scaled to fit different resolutions. Can be
        // specified using percentage units.
        width: 1050,

        height: 700,

        // Factor of the display size that should remain empty around the content
        margin: 0.1,

        math: {
          mathjax: 'https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.9/MathJax.js',
          config: 'TeX-AMS_HTML-full',
          tex2jax: {
            inlineMath: [['\\(','\\)']],
            displayMath: [['\\[','\\]']],
            balanceBraces: true,
            processEscapes: false,
            processRefs: true,
            processEnvironments: true,
            preview: 'TeX',
            skipTags: ['script','noscript','style','textarea','pre','code'],
            ignoreClass: 'tex2jax_ignore',
            processClass: 'tex2jax_process'
          },
        },

        // reveal.js plugins
        plugins: [QuartoLineHighlight, PdfExport, RevealMenu, RevealChalkboard, QuartoSupport,

          RevealMath,
          RevealNotes,
          RevealSearch,
          RevealZoom
        ]
      });
    </script>
    
    <script>
      // htmlwidgets need to know to resize themselves when slides are shown/hidden.
      // Fire the "slideenter" event (handled by htmlwidgets.js) when the current
      // slide changes (different for each slide format).
      (function () {
        // dispatch for htmlwidgets
        function fireSlideEnter() {
          const event = window.document.createEvent("Event");
          event.initEvent("slideenter", true, true);
          window.document.dispatchEvent(event);
        }

        function fireSlideChanged(previousSlide, currentSlide) {
          fireSlideEnter();

          // dispatch for shiny
          if (window.jQuery) {
            if (previousSlide) {
              window.jQuery(previousSlide).trigger("hidden");
            }
            if (currentSlide) {
              window.jQuery(currentSlide).trigger("shown");
            }
          }
        }

        // hookup for slidy
        if (window.w3c_slidy) {
          window.w3c_slidy.add_observer(function (slide_num) {
            // slide_num starts at position 1
            fireSlideChanged(null, w3c_slidy.slides[slide_num - 1]);
          });
        }

      })();
    </script>

    <script id="quarto-html-after-body" type="application/javascript">
    window.document.addEventListener("DOMContentLoaded", function (event) {
      const toggleBodyColorMode = (bsSheetEl) => {
        const mode = bsSheetEl.getAttribute("data-mode");
        const bodyEl = window.document.querySelector("body");
        if (mode === "dark") {
          bodyEl.classList.add("quarto-dark");
          bodyEl.classList.remove("quarto-light");
        } else {
          bodyEl.classList.add("quarto-light");
          bodyEl.classList.remove("quarto-dark");
        }
      }
      const toggleBodyColorPrimary = () => {
        const bsSheetEl = window.document.querySelector("link#quarto-bootstrap");
        if (bsSheetEl) {
          toggleBodyColorMode(bsSheetEl);
        }
      }
      toggleBodyColorPrimary();  
      const tabsets =  window.document.querySelectorAll(".panel-tabset-tabby")
      tabsets.forEach(function(tabset) {
        const tabby = new Tabby('#' + tabset.id);
      });
      const isCodeAnnotation = (el) => {
        for (const clz of el.classList) {
          if (clz.startsWith('code-annotation-')) {                     
            return true;
          }
        }
        return false;
      }
      const onCopySuccess = function(e) {
        // button target
        const button = e.trigger;
        // don't keep focus
        button.blur();
        // flash "checked"
        button.classList.add('code-copy-button-checked');
        var currentTitle = button.getAttribute("title");
        button.setAttribute("title", "Copied!");
        let tooltip;
        if (window.bootstrap) {
          button.setAttribute("data-bs-toggle", "tooltip");
          button.setAttribute("data-bs-placement", "left");
          button.setAttribute("data-bs-title", "Copied!");
          tooltip = new bootstrap.Tooltip(button, 
            { trigger: "manual", 
              customClass: "code-copy-button-tooltip",
              offset: [0, -8]});
          tooltip.show();    
        }
        setTimeout(function() {
          if (tooltip) {
            tooltip.hide();
            button.removeAttribute("data-bs-title");
            button.removeAttribute("data-bs-toggle");
            button.removeAttribute("data-bs-placement");
          }
          button.setAttribute("title", currentTitle);
          button.classList.remove('code-copy-button-checked');
        }, 1000);
        // clear code selection
        e.clearSelection();
      }
      const getTextToCopy = function(trigger) {
          const codeEl = trigger.previousElementSibling.cloneNode(true);
          for (const childEl of codeEl.children) {
            if (isCodeAnnotation(childEl)) {
              childEl.remove();
            }
          }
          return codeEl.innerText;
      }
      const clipboard = new window.ClipboardJS('.code-copy-button:not([data-in-quarto-modal])', {
        text: getTextToCopy
      });
      clipboard.on('success', onCopySuccess);
      if (window.document.getElementById('quarto-embedded-source-code-modal')) {
        const clipboardModal = new window.ClipboardJS('.code-copy-button[data-in-quarto-modal]', {
          text: getTextToCopy,
          container: window.document.getElementById('quarto-embedded-source-code-modal')
        });
        clipboardModal.on('success', onCopySuccess);
      }
        var localhostRegex = new RegExp(/^(?:http|https):\/\/localhost\:?[0-9]*\//);
        var mailtoRegex = new RegExp(/^mailto:/);
          var filterRegex = new RegExp("https:\/\/cct-datascience\.github\.io\/ggplot2-workshop-series\/");
        var isInternal = (href) => {
            return filterRegex.test(href) || localhostRegex.test(href) || mailtoRegex.test(href);
        }
        // Inspect non-navigation links and adorn them if external
     	var links = window.document.querySelectorAll('a[href]:not(.nav-link):not(.navbar-brand):not(.toc-action):not(.sidebar-link):not(.sidebar-item-toggle):not(.pagination-link):not(.no-external):not([aria-hidden]):not(.dropdown-item):not(.quarto-navigation-tool):not(.about-link)');
        for (var i=0; i<links.length; i++) {
          const link = links[i];
          if (!isInternal(link.href)) {
            // undo the damage that might have been done by quarto-nav.js in the case of
            // links that we want to consider external
            if (link.dataset.originalHref !== undefined) {
              link.href = link.dataset.originalHref;
            }
              // target, if specified
              link.setAttribute("target", "_blank");
              if (link.getAttribute("rel") === null) {
                link.setAttribute("rel", "noopener");
              }
              // default icon
              link.classList.add("external");
          }
        }
      function tippyHover(el, contentFn, onTriggerFn, onUntriggerFn) {
        const config = {
          allowHTML: true,
          maxWidth: 500,
          delay: 100,
          arrow: false,
          appendTo: function(el) {
              return el.closest('section.slide') || el.parentElement;
          },
          interactive: true,
          interactiveBorder: 10,
          theme: 'light-border',
          placement: 'bottom-start',
        };
        if (contentFn) {
          config.content = contentFn;
        }
        if (onTriggerFn) {
          config.onTrigger = onTriggerFn;
        }
        if (onUntriggerFn) {
          config.onUntrigger = onUntriggerFn;
        }
          config['offset'] = [0,0];
          config['maxWidth'] = 700;
        window.tippy(el, config); 
      }
      const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
      for (var i=0; i<noterefs.length; i++) {
        const ref = noterefs[i];
        tippyHover(ref, function() {
          // use id or data attribute instead here
          let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
          try { href = new URL(href).hash; } catch {}
          const id = href.replace(/^#\/?/, "");
          const note = window.document.getElementById(id);
          if (note) {
            return note.innerHTML;
          } else {
            return "";
          }
        });
      }
      const findCites = (el) => {
        const parentEl = el.parentElement;
        if (parentEl) {
          const cites = parentEl.dataset.cites;
          if (cites) {
            return {
              el,
              cites: cites.split(' ')
            };
          } else {
            return findCites(el.parentElement)
          }
        } else {
          return undefined;
        }
      };
      var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
      for (var i=0; i<bibliorefs.length; i++) {
        const ref = bibliorefs[i];
        const citeInfo = findCites(ref);
        if (citeInfo) {
          tippyHover(citeInfo.el, function() {
            var popup = window.document.createElement('div');
            citeInfo.cites.forEach(function(cite) {
              var citeDiv = window.document.createElement('div');
              citeDiv.classList.add('hanging-indent');
              citeDiv.classList.add('csl-entry');
              var biblioDiv = window.document.getElementById('ref-' + cite);
              if (biblioDiv) {
                citeDiv.innerHTML = biblioDiv.innerHTML;
              }
              popup.appendChild(citeDiv);
            });
            return popup.innerHTML;
          });
        }
      }
    });
    </script>
    

</body></html>